src/watchdog.rs

```diff
@@ -0,0 +1,108 @@
+//! API for the IWDG
+//!
+//! You can activate the watchdog by calling `start` or the setting appropriate
+//! device option bit when programming.
+//!
+//! After activating the watchdog, you'll have to regularly `feed` the watchdog.
+//! If more time than `timeout` has gone by since the last `feed`, your
+//! microcontroller will be reset.
+//!
+//! This is useful if you fear that your program may get stuck. In that case it
+//! won't feed the watchdog anymore, the watchdog will reset the microcontroller
+//! and thus your program will function again.
+//!
+//! **Attention**:
+//!
+//! The IWDG runs on a separate 40kHz low-accuracy clock (30kHz-60kHz). You may
+//! want to some buffer in your interval.
+//!
+//! # Example
+//! ``` no_run
+//! use stm32f0xx_hal as hal;
+//!
+//! use crate::hal::stm32;
+//! use crate::hal::prelude::*;
+//! use crate::hal:watchdog::Watchdog;
+//! use crate::hal:time::Hertz;
+//!
+//! let mut p = stm32::Peripherals::take().unwrap();
+//!
+//! let mut iwdg = Watchdog::new(p.iwdg);
+//! iwdg.start(Hertz(100));
+//! loop {}
+//! // Whoops, got stuck, the watchdog issues a reset after 10 ms
+//! iwdg.feed();
+//! ```
+use embedded_hal::watchdog;
+
+use crate::stm32::IWDG;
+use crate::time::Hertz;
+
+/// Watchdog instance
+pub struct Watchdog {
+    iwdg: IWDG,
+}
+
+impl watchdog::Watchdog for Watchdog {
+    /// Feed the watchdog, so that at least one `period` goes by before the next
+    /// reset
+    fn feed(&mut self) {
+        self.iwdg.kr.write(|w| w.key().reset());
+    }
+}
+
+/// Timeout configuration for the IWDG
+#[derive(PartialEq, PartialOrd, Clone, Copy)]
+pub struct IwdgTimeout {
+    psc: u8,
+    reload: u16,
+}
+
+impl Into<IwdgTimeout> for Hertz {
+    /// This converts the value so it's usable by the IWDG
+    /// Due to conversion losses, the specified frequency is a maximum
+    ///
+    /// It can also only represent values < 10000 Hertz
+    fn into(self) -> IwdgTimeout {
+        let mut time = 40_000 / 4 / self.0;
+        let mut psc = 0;
+        let mut reload = 0;
+        while psc < 7 {
+            reload = time;
+            if reload < 0x1000 {
+                break;
+            }
+            psc += 1;
+            time /= 2;
+        }
+        // As we get an integer value, reload is always below 0xFFF
+        let reload = reload as u16;
+        IwdgTimeout { psc, reload }
+    }
+}
+impl Watchdog {
+    pub fn new(iwdg: IWDG) -> Self {
+        Self { iwdg }
+    }
+}
+
+impl watchdog::WatchdogEnable for Watchdog {
+    type Time = IwdgTimeout;
+    fn start<T>(&mut self, period: T)
+    where
+        T: Into<IwdgTimeout>,
+    {
+        let time: IwdgTimeout = period.into();
+        // First kick the watchdog
```

Comment:
Please find a different phrase than "kick the watchdog" in the comments. I know it gets used a lot (and I've certainly done it in the past) but some people find that offensive.

```
/// Feeds an existing watchdog to ensure the processor isn't reset. Sometimes
/// commonly referred to as "kicking" or "refreshing".
#[cfg(feature = "unproven")]
pub trait Watchdog {
    /// Triggers the watchdog. This must be done once the watchdog is started
    /// to prevent the processor being reset.
    fn feed(&mut self);
}
```